### PR TITLE
 DASHBUILDE-252: CMS perspective UI redesign

### PR DESCRIPTION
**Pull request by dgutierr**
Brand new UXD design of the left sidebar in the Content Management perspective.

See video => https://issues.jboss.org/secure/attachment/12425542/dashbuilder_cms_uxdreview.mp4

@ederign @jhrcek Can you take a look please?

### CODE REVIEW

**Comment by jhrcek** [comment]
Jenkins execute full downstream build

File: dashbuilder-client/dashbuilder-navigation-client/src/main/resources/org/dashbuilder/client/navigation/resources/i18n/NavigationConstants.properties [hunk 1]
```diff
@@ -14,20 +14,25 @@
...
+perspectiveIconTitle=A hard-coded static perspective
 moveUp=Move up
 moveDown=Move down
 moveFirst=Move first
 moveLast=Move last
 saveChanges=Save changes?
 save=Save
 cancel=Cancel
-navItemNotFound="Not found: {0}
+navGroupNotFound=The navigation group configured might have been deleted or does not exit. Please, check this component configuration.
```
Comment (jhrcek) [inline dashbuilder-client/dashbuilder-navigation-client/src/main/resources/org/dashbuilder/client/navigation/resources/i18n/NavigationConstants.properties]: check this component configuration -> check configuration of this component

**Comment by kkufova** [comment]
I like the changes a lot! It is much easier to work with perspectives now when both **Pages** and **Navigation** are in the same panel and the user doesn't need to switch between them. Creating new navigation trees is more user-friendly and I like a lot that it is not possible to click on an empty custom tree and expand it. Overall, I think this works and looks really good.

Found issues:
-------------------------------------------------------------------------------
CREATING CUSTOM NAVIGATION TREES: I discovered that when creating custom navigation trees or doing any changes in the **Navigation** panel and saving them continuously, if I once click **Cancel** in the **Save changes?** dialog, _all_ the changes disappear and the panel goes back to its very original state. This needs to be fixed so that only the changes that were made _after_ clicking **Save** for the last time disappear.

I don't know if this is an example of an infinite loop, but I managed to do something similar. If you create two pages with for example the Menu Bar Component, which has the same Navigation Group, it allows you to open the menu infinitely. See the video:
http://file.brq.redhat.com/~kkufova/4.10.2017/screencast-10-04-2017-10-57-38-AM.webm
Do we want to handle cases like this?

I noticed that when an infinite loop happens, the reason in the warning message is no longer displayed. Is this on purpose? If so, the word **Reason:** should be probably removed. :-)

When you want to pass an empty tree as a Navigation Group when adding a Tile Navigator Component, a message _Please, make sure the group configured is not empty_ appears. I think this is great, but it only works like this for the tiles; for all the other components (Tab List Component, Tree Navigator Component, Menu Bar Component) the component is created anyway and states _No items found_. Could we unify this behavior?

Notes:
-------------------------------------------------------------------------------
It happened to me that the dialog **Save changes?** disappeared and never showed up again even though I was doing some changes and that it was only possible to add one random perspective into the tree. Also, that the Tile Navigator Component didn't work at all for custom trees (nothing was displayed) and worked badly for the default Top Menu tree (breadcrumbs were accumulating but nothing was displayed). However, I don't know how to reproduce any of these issues, as it fixed itself after I did some clicking (and accidentally deleted all my work). Beware of this behavior. :-)

File: dashbuilder-client/dashbuilder-navigation-client/src/main/resources/org/dashbuilder/client/navigation/resources/i18n/NavigationConstants.properties (hunk 1 above)
Comment (dgutierr) [reply dashbuilder-client/dashbuilder-navigation-client/src/main/resources/org/dashbuilder/client/navigation/resources/i18n/NavigationConstants.properties]: Done

**Comment by dgutierr** [comment]
@jhrcek @kkufova I've applied the changes suggested above.

1. "The changes that were made after clicking Save for the last time disappear"

Fixed.

2. False infinite loop. Do we want to handle cases like this?

This is not a real infinite loop as the application never goes out of control. This is a kind of weird configuration scenario which I'm not sure it might be considered an error.

3. "Please, make sure the group configured is not empty".  Could we unify this behavior?

Done.

4. It happened to me that the dialog Save changes? disappeared and never showed up again.

I didn't happened to me. But please, keep trying to reproduce it.

5. Collapse of the "Navigation" panel.

Done.

Here is an updated video with those changes => https://issues.jboss.org/secure/attachment/12425666/dashbuilder_cms_uxdreview2.mp4

**Comment by kkufova** [comment]
@dgutierr 
Thank you for the fixes! I really like that now it is possible to collapse the panels by clicking on the panel names. I tried all the scenarios that didn't work before and everything is OK.

I have a few questions and suggestions, could you please let me know what you think?

Regarding the naming of individual pages; I named my custom pages _Page no. 1_, _Page no. 2_, _Page no. 3_ and _Page no. 4_, but when I want to add them to some tree in the **Navigation** panel, only the numbers are displayed:
screenshot from 2017-10-05 14-34-38
Does this happen because the pages are incorrectly named, or is this a bug? If there are some restrictions on the page names, I think it would be best to not let the user to name the page badly when creating it.

After clicking **Cancel** in the **Save changes?** dialog, would it be possible to remember the state of the trees (collapsed or not)? Now each tree closes and the user has to open everything again.

Would it be possible to remove a page from the list of pages to be added to a tree once it has been added to a particular tree? I use the same page in one tree multiple times for the sake of testing, but I cannot think of a scenario in the real world. It would also shorten the list of available pages when creating a complex tree. WDYT?

Why is it possible to add a _divider_ to a tree? It never shows in any navigation component, only in the tree itself in the **Navigation** panel. Could we maybe remove the dividers completely, if there is no use for them?

If you want me to create a separate Jira for any of this, please let me know.

**Comment by dgutierr** [comment]
@kkufova 

1.- The page names seems to be a bug => JIRA

2.- Clicking on Cancel collapses everything . It is a bit of extra work to save the expand status for each node but I agree it is worth to give it a try.

3. A page can be reused as a template in different scenarios. I would not limit the possibilities the user has here. 

4. The nav tree configuration including its dividers is decoupled from the nav widgets. It is up to the widget to display the items under its configured nav group. The thing is that we do not know during the creation of our custom trees which are going to be the widgets using them. As you suggest, we could remove completely the dividers. Only the menu bar widget will be affected as it is the only one supporting them.

**Comment by kkufova** [comment]
@dgutierr 
Alright, thanks! Have you already implemented any of the mentioned changes? There is a failing test (`NavTreeEditorTest.testNewDivider`), so I cannot get the newest `.war` file to check it.

**Comment by dgutierr** [comment]
@kkufova I fixed the newDivider test, my fault sorry.

The latest changes suggested (see 1, 2 above) are not implemented yet.

**Comment by kkufova** [comment]
@dgutierr I tried to check how it works without the dividers but they are still there. Do you want me to create two more separate Jiras (one for the dividers and one for the collapsing) or will we do it as a part of this PR?

**Comment by dgutierr** [comment]
@kkufova The dividers should work as usual. They have not been removed nor I plan to do so. I'd like to get more feedback from UXD about it. 

As for the collapsing enhancement, I'm working on it. No JIRA needed.

**Comment by kkufova** [comment]
@dgutierr, my fault, I am sorry! Now I see how the dividers are displayed in the Menu Bar Component, I didn't notice it before. I completely agree that we should either leave them as they are, or discuss it with the UX team. I cannot tell which changes in the PR were made at the latest, so I thought based on your message that you were going to remove the dividers. :-)

**Comment by dgutierr** [comment]
@kkufova Both outstanding issues fixed.

1.- DASHBUILDE-255 
2.- Clicking on Cancel collapses everything

Please, take a look.

**Comment by kkufova** [comment]
@dgutierr 
Thanks, works great! 

I discovered one more issue with renaming pages. If you create a page (called for example `test`) with a navigation component and then rename it (i.e. to `myPage`):
1. a copy of the page is created below the original page. See:
page_copy_bug
2. Then, if you want to add this page (`myPage`) to a custom tree, the page is displayed under the name _before_ it was renamed (`test`) and it doesn't offer the selection of a navigation group as it should.
3. For some reason, the original page then appears again in the **Pages** panel and cannot be deleted, so the page is now in the panel twice (`test` and `myPage`).

------

Also, if you delete a page that is already used in some custom tree, would it be possible to display some notification or warning that some tree is going to be affected, or remove the page automatically from all trees?

------

And one more question: could you please explain to me one more time why is the navigation group setting while creating trees needed, when it only overrides the original setting of the page? Now when we have both **Pages** and **Navigation** in one place, wouldn't it be enough to be able to set the navigation group only when editing some page and not while editing a tree? I think it would significantly simplify the process unless it offers some additional functionality.

Please let me know if you want me to create Jiras for any of this. These issues do not block this PR.

**Comment by dgutierr** [comment]
@kkufova You are right. The renaming of a page is not working well and it is also causing some side effects in the navigation bahaviour. Can you please file a JIRA?

As for page deletion, I've pushed a minor fix to ensure those items linked to a deleted page are removed from the tree editor. 

Regarding the nav group setting. The nav item's group (from the global nav tree configuration) is optional. When set, it will overwrite the static nav group configured at the component level. Why is this setting helpful? Basically is a powerful mechanism for allowing the reuse of the same perspective using different navigation entries. As you can see in the https://issues.jboss.org/browse/DASHBUILDE-246 videos, the entries displayed in the nav components are tied to the nav group configured in the nav items. Without this setting you are forced to create different copies of the same perspective if you want to have different perspectives that only differ in the nav entries their nav components show.

Hope it helps.

**Comment by kkufova** [comment]
@dgutierr 
Thanks, the deleting pages works really nice. I created a Jira for the renaming ([DASHBUILDE-256](https://issues.jboss.org/browse/DASHBUILDE-256)).

**Comment by dgutierr** [comment]
@ederign Can you please take a last look at this PR? QE already approve it so I'd like to have it merged before EOD. ThanXs mate!

**Comment by kkufova** [comment]
@dgutierr 
Regarding the [DASHBUILDE-257](https://issues.jboss.org/browse/DASHBUILDE-257) fix: it works in the beginning, but if you:

1. first create a custom tree and
2. then create a very first page

and want to add the page to the tree, there is no option **+ New Page**. You need to create a _new_ tree first for the option to be there.
BUT, if you then delete the page (so there is no page left again), the **+ New Page** option is still there and it doesn't work properly; it shows some perspective called "1" (which cannot be added to the tree) and _Perspective not found_. Could you please have a look at this?

And could we please handle new bugs in separate PRs next time?

**Comment by dgutierr** [comment]
@kkufova I've finished the "+ New page" issue.  Please take a look, everything should work ok now.

**Comment by kkufova** [comment]
@dgutierr 
It works fine, thanks!

**Comment by dgutierr** [comment]
@kkufova is it ready to approve?

**Pull request merged**